### ISSUE TITLE
Document the best practices for middleware development

### ISSUE THREAD

**Issue by bbatsov** [nrepl/nrepl#144]
That's mostly a reminder for me to add some section in the documentation about this. (but I wouldn't mind if someone else wants to tackle this)

I've noticed that often people design middlewares in weird ways, which leads me to believe we have to documented the best practices accordingly.

**Comment by kolharsam** [comment]
@bbatsov can I help with this? I did check the nrepl website and there was no explicit section with details about this.

**Comment by bbatsov** [comment]
Sure!

**Comment by kolharsam** [comment]
@bbatsov how do I get started?

**Comment by bbatsov** [comment]
I've just committed some basic outline of the page that you can work with. I'm thinking we should walk people over the creation of some simple middleware and cover suggestions for the code structure, error handling, writing tests, request/response format, middleware descriptors, interaction between different middlewares, and so on.

**Comment by shen-tian** [comment]
@bbatsov: do you have any notes on what are the important bits are under each topic?

I've been looking at how we test middleware. As it stands, `core_test` does a lot of heavy lifting, and it's all in the form of integration tests. 

I tried to refactor out some of these tests into single-middleware unit tests, and it's looking promising:

https://github.com/shen-tian/nrepl/blob/print-tests/test/clojure/nrepl/middleware/print_test.clj

**Comment by bbatsov** [comment]
Love it! I've long meant to do something like this myself, as I hate convoluted tests.

> @bbatsov: do you have any notes on what are the important bits are under each topic?

No real notes, just ideas in my head:

* showing how to interact with middleware like `session`
* making sure you're returning some meaningful statuses to clients
* separating the business logic out of the middleware, so it's easier to test it and you'd have only a few integration tests for the middleware
* naming conventions for middleware and ops
* dealing with the limitations of bencode (e.g. you can have boolean flags in requests)
* configuration techniques (e.g. dynamic vars, options hashes)
* documenting middleware
* creating middleware that just modifies requests/responses (as opposed to the classic variant - a middleware that responds directly to requests)
* avoiding middlewares that handle a ton of unrelated ops (e.g. that's what `refactor-nrepl` does - one middleware for everything)

Things along those lines.

**Comment by kolharsam** [comment]
@bbatsov cool! On the [docs](https://nrepl.org/nrepl/0.7.0/design/middleware.html) you've mentioned that all default functionality exposed by nrepl is through middleware. So why don't we go over one of the simpler ones and have described along the pointers that you've given above?

**Comment by bbatsov** [comment]
Yeah, that's an option. I think the print middleware is also quite interesting, as it just modifies the behavior of `eval` (and related). I was thinking another good example might be some "benchmark" middleware which just tracks how much it took to complete a request and adds that to the response. That's actually a somewhat useful idea for improving nREPL itself. :-)

**Comment by kolharsam** [comment]
The "benchmark" middleware seems like a useful idea. Could that just be a part of the `print` middleware? 

It definitely seems like an interesting problem to be working on.
